packages/preferences/src/browser/views/components/preference-number-input.tsx

```diff
@@ -17,41 +17,86 @@
 
 import * as React from 'react';
 import { Preference } from '../../util/preference-types';
+import { Event } from '@theia/core/lib/common';
 
 interface PreferenceNumberInputProps {
     preferenceDisplayNode: Preference.NodeWithValueInSingleScope;
     setPreference(preferenceName: string, preferenceValue: number): void;
+    onPreferenceReset: Event<string>;
 }
 
-export const PreferenceNumberInput: React.FC<PreferenceNumberInputProps> = ({ preferenceDisplayNode, setPreference }) => {
+export const PreferenceNumberInput: React.FC<PreferenceNumberInputProps> = ({ preferenceDisplayNode, setPreference, onPreferenceReset }) => {
     const { id } = preferenceDisplayNode;
     const { data, value } = preferenceDisplayNode.preference;
 
     const externalValue = (value !== undefined ? value : data.defaultValue) || '';
 
     const [currentTimeout, setCurrentTimetout] = React.useState<number>(0);
     const [currentValue, setCurrentValue] = React.useState<string>(externalValue);
+    const [currentMessage, setCurrentMessage] = React.useState<string>('');
+    const [isErrorMessageVisible, setErrorMessageVisibility] = React.useState<boolean>(false);
 
     React.useEffect(() => {
+        let mounted: boolean = true;
         setCurrentValue(externalValue);
+        onPreferenceReset(preferenceId => {
+            if (mounted && preferenceId === id) {
+                setCurrentValue(data.defaultValue);
+                setCurrentMessage('');
+            }
+        });
+        return function cleanup(): void {
+            mounted = false;
+        };
     }, [externalValue]);
 
     const onChange = React.useCallback(e => {
-        const { value: newValue } = e.target;
         clearTimeout(currentTimeout);
-        const newTimeout = setTimeout(() => setPreference(id, Number(newValue)), 750);
-        setCurrentTimetout(Number(newTimeout));
+        const { value: newValue } = e.target;
         setCurrentValue(newValue);
+        const preferenceValue: number = Number(newValue);
+        const { isValid, message } = getErrorMessage(preferenceValue);
+        setCurrentMessage(message);
+        setErrorMessageVisibility(!isValid);
+        if (isValid) {
+            const newTimeout = setTimeout(() => setPreference(id, preferenceValue), 750);
+            setCurrentTimetout(Number(newTimeout));
+        }
     }, [currentTimeout]);
 
+    const onBlur = () => setErrorMessageVisibility(false);
+    const onFocus = () => setErrorMessageVisibility(!!currentMessage.length);
```

Comment:
You may want to use `React.useCallback` for these. Using plain arrow functions can be detrimental to performance because they always force a rerender of whatever component they're used as a prop for.

Reply:
I have updated the code to use `React.useCallback` for both `onFocus` and `onBlur`

Reply:
I think a typo cropped up in your `onFocus` callback, but otherwise this change looks good. Another way to control visibility on focus would be to use CSS and the [`:focus-within`](https://developer.mozilla.org/en-US/docs/Web/CSS/:focus-within). That would require a little less state handling inside the component, but either way is fine with me.